### ISSUE TITLE
Shared memory errors occur when sending large JPEG images as bytestrings asynchronously from BLS

### ISSUE THREAD

**Issue by Bazer47** [triton-inference-server/server#6633]
Hi Maintainers & Contributors,

I would like to ask following questions.

**Description**
We are using currently [BLS from Python backend](https://github.com/triton-inference-server/python_backend/blob/main/examples/bls/README.md) to send multiple (2-3) asynchronous requests to different ensemble models (AI model running on onnxruntime with pre/post-processing). Each request contains encoded JPEG in bytestring (dtype TYPE_STRING), because each model has some specific preprocessing and the images have various shape. The input to the BLS is also a bytestring with JPEG image and some boolean array with a model selection.

It seems that there is a big difference if we send a JPEG image of size 1MB and if the size is 1.8MB. When we send the larger image there are usually some memory problems:

Client error
> tritonclient.utils.InferenceServerException: Failed to process the request(s) for model instance 'BLS_model_0_0', message: TritonModelException: in ensemble 'cv_model_1', Failed to process the request(s) for model instance 'cv_model_1_preprocessing_0_0', message: error: unpack_from requires a buffer of at least 2066633 bytes for unpacking 2066629 bytes at offset 4 (actual buffer size is 5888)
> 
> At:
>   /opt/tritonserver/backends/python/triton_python_backend_utils.py(117): deserialize_bytes_tensor

Docker log when the container suddenly stops
> terminate called after throwing an instance of 'boost::interprocess::bad_alloc'
>   what():  boost::interprocess::bad_alloc
> Signal (6) received.
> 
> 0# 0x0000557C7BF5383D in tritonserver
>  1# 0x00007F92185AD520 in /lib/x86_64-linux-gnu/libc.so.6
>  2# pthread_kill in /lib/x86_64-linux-gnu/libc.so.6
>  3# raise in /lib/x86_64-linux-gnu/libc.so.6
>  4# abort in /lib/x86_64-linux-gnu/libc.so.6
>  5# 0x00007F921891CB9E in /lib/x86_64-linux-gnu/libstdc++.so.6
>  6# 0x00007F921892820C in /lib/x86_64-linux-gnu/libstdc++.so.6
>  7# 0x00007F9218928277 in /lib/x86_64-linux-gnu/libstdc++.so.6
>  8# 0x00007F92189284D8 in /lib/x86_64-linux-gnu/libstdc++.so.6
>  9# 0x00007F921525F4BD in /opt/tritonserver/backends/python/libtriton_python.so
> 10# 0x00007F92152BAA34 in /opt/tritonserver/backends/python/libtriton_python.so
> 11# 0x00007F921527482E in /opt/tritonserver/backends/python/libtriton_python.so
> 12# 0x00007F9215287D13 in /opt/tritonserver/backends/python/libtriton_python.so
> 13# 0x00007F921528CFD8 in /opt/tritonserver/backends/python/libtriton_python.so
> 14# 0x00007F92152931C6 in /opt/tritonserver/backends/python/libtriton_python.so
> 15# TRITONBACKEND_ModelInstanceExecute in /opt/tritonserver/backends/python/libtriton_python.so
> 16# 0x00007F9218F91FE4 in /opt/tritonserver/bin/../lib/libtritonserver.so
> 17# 0x00007F9218F9234B in /opt/tritonserver/bin/../lib/libtritonserver.so
> 18# 0x00007F92190658BD in /opt/tritonserver/bin/../lib/libtritonserver.so
> 19# 0x00007F9218F96774 in /opt/tritonserver/bin/../lib/libtritonserver.so
> 20# 0x00007F9218956253 in /lib/x86_64-linux-gnu/libstdc++.so.6
> 21# 0x00007F92185FFAC3 in /lib/x86_64-linux-gnu/libc.so.6
> 22# clone in /lib/x86_64-linux-gnu/libc.so.6
> 
> Signal (11) received.
>  0# 0x0000557C7BF5383D in tritonserver
>  1# 0x00007F92185AD520 in /lib/x86_64-linux-gnu/libc.so.6
>  2# abort in /lib/x86_64-linux-gnu/libc.so.6
>  3# 0x00007F921891CB9E in /lib/x86_64-linux-gnu/libstdc++.so.6
>  4# 0x00007F921892820C in /lib/x86_64-linux-gnu/libstdc++.so.6
>  5# 0x00007F9218928277 in /lib/x86_64-linux-gnu/libstdc++.so.6
>  6# 0x00007F92189284D8 in /lib/x86_64-linux-gnu/libstdc++.so.6
>  7# 0x00007F921525F4BD in /opt/tritonserver/backends/python/libtriton_python.so
>  8# 0x00007F92152BAA34 in /opt/tritonserver/backends/python/libtriton_python.so
>  9# 0x00007F921527482E in /opt/tritonserver/backends/python/libtriton_python.so
> 10# 0x00007F9215287D13 in /opt/tritonserver/backends/python/libtriton_python.so
> 11# 0x00007F921528CFD8 in /opt/tritonserver/backends/python/libtriton_python.so
> 12# 0x00007F92152931C6 in /opt/tritonserver/backends/python/libtriton_python.so
> 13# TRITONBACKEND_ModelInstanceExecute in /opt/tritonserver/backends/python/libtriton_python.so
> 14# 0x00007F9218F91FE4 in /opt/tritonserver/bin/../lib/libtritonserver.so
> 15# 0x00007F9218F9234B in /opt/tritonserver/bin/../lib/libtritonserver.so
> 16# 0x00007F92190658BD in /opt/tritonserver/bin/../lib/libtritonserver.so
> 17# 0x00007F9218F96774 in /opt/tritonserver/bin/../lib/libtritonserver.so
> 18# 0x00007F9218956253 in /lib/x86_64-linux-gnu/libstdc++.so.6
> 19# 0x00007F92185FFAC3 in /lib/x86_64-linux-gnu/libc.so.6
> 20# clone in /lib/x86_64-linux-gnu/libc.so.6

Sometimes it has problems with shared memory pool size and shared memory regions
> tritonclient.utils.InferenceServerException: Failed to process the request(s) for model instance 'BLS_model_0_0', message: TritonModelException: in ensemble 'cv_model_1', Failed to increase the shared memory pool size for key 'triton_python_backend_shm_region_18' to 3145728 bytes. If you are running Triton inside docker, use '--shm-size' flag to control the shared memory region size. Error: Invalid argument
> 
> At:
>   /model_repository/BLS_model/1/model.py(178): execute

We also tested scaling down the images before sending them to the models and **it helped**.
We tested `--shm-size` argument from `3g` to `10g` and it did not help.

**Triton Information**
We used standard NGC Docker image `nvcr.io/nvidia/tritonserver:23.09-py3` (the same problem is mentioned in this issue #6513 ). But we are currently using custom build only cpu Docker image 23.12 (this helped us shifting the issue from some GPU memory errors to this shared memory errors, easier debugging).

**Questions**
Is there a way how to increase the memory in order to send asynchronously any reasonable image without any error? What is the best practice for sending asynchronously images from BLS to each model? How to optimize this process?

**TL;DR**
Shared memory errors occur when sending large JPEG images (1-2MB) as bytestrings asynchronously from BLS to several models. Scaling them down in the BLS helps. How to optimize it? How to increase the memory needed to send the requests with these images? 

Thank you!

**Comment by kthui** [comment]
This could be a bug since increasing the `--shm-size` did not help. I have filed a ticket for us to investigate further.

**Comment by kthui** [comment]
Hi @Bazer47, I am not able to replicate the issue on either 23.09 or on the main branch. Would you be able to try the latest release and see if the issue still persisted?

**Comment by kthui** [comment]
We are closing this issue due to lack of activity. Please feel free to file anther issue if the problem persisted.
